pkg/controller/storagecluster/volumesnapshotterclasses.go

```diff
@@ -0,0 +1,113 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	snapapi "github.com/kubernetes-csi/external-snapshotter/v2/pkg/apis/volumesnapshot/v1beta1"
+	ocsv1 "github.com/openshift/ocs-operator/pkg/apis/ocs/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type SnapShotterType string
+
+const (
+	rbdSnapShotter    SnapShotterType = "rbd"
+	cephfsSnapShotter SnapShotterType = "cephfs"
+)
+
+// secret name and namespace for snapshotter class
+const (
+	snapshotterSecretName      = "csi.storage.k8s.io/snapshotter-secret-name"
+	snapshotterSecretNamespace = "csi.storage.k8s.io/snapshotter-secret-namespace"
+	snapshotterNamePrefix      = "snapshotNamePrefix"
+)
+
+var (
+	snapshotterSecretNameMap = map[SnapShotterType]string{
+		rbdSnapShotter:    "rook-csi-rbd-provisioner",
+		cephfsSnapShotter: "csi-cephfs-secret",
+	}
+)
+
+// newVolumeSnapshotClass returns a new VolumeSnapshotter class backed by provided snapshotter type
+// available 'snapShotterType' values are 'rbd' and 'cephfs'
+func newVolumeSnapshotClass(instance *ocsv1.StorageCluster, snapShotterType SnapShotterType) *snapapi.VolumeSnapshotClass {
+	retSC := &snapapi.VolumeSnapshotClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: generateNameForSnapshotClass(instance, snapShotterType),
+		},
+		Driver: generateNameForSnapshotClassDriver(instance, snapShotterType),
+		Parameters: map[string]string{
+			"clusterID":                instance.Namespace,
+			snapshotterSecretName:      snapshotterSecretNameMap[snapShotterType],
+			snapshotterSecretNamespace: instance.Namespace,
+		},
+		DeletionPolicy: snapapi.VolumeSnapshotContentDelete,
+	}
+	return retSC
+}
+
+func newSnapshotClasses(instance *ocsv1.StorageCluster) []*snapapi.VolumeSnapshotClass {
+	scs := []*snapapi.VolumeSnapshotClass{
+		newVolumeSnapshotClass(instance, rbdSnapShotter),
+		newVolumeSnapshotClass(instance, cephfsSnapShotter),
+	}
+	return scs
+}
+
+func (r *ReconcileStorageCluster) createSnapshotClasses(vscs []*snapapi.VolumeSnapshotClass, reqLogger logr.Logger) error {
+	for _, vsc := range vscs {
+		existing := &snapapi.VolumeSnapshotClass{}
+		err := r.client.Get(context.TODO(), types.NamespacedName{Name: vsc.Name, Namespace: vsc.Namespace}, existing)
+		if err != nil {
```

Comment:
@aruniiird from, 68-82 lines can be combined.. 
`if err != nil && if errors.IsNotFound(err) { create -> return check ->continue/}` -> else {...}
Also you can use different errors to avoid overwriting`err` var in a scope .
For ex: for deep inside err, capture it on "createErr" instead of overwriting `err` var.

Reply:
Please follow the pattern here: https://github.com/openshift/ocs-operator/blob/master/pkg/controller/storagecluster/storageclasses.go#L39-L71

Reply:
@jarrpa , I have removed the funneling of errors, but I didn't feel the final `else` very compelling in the above example you quoted. Added a `continue` to go on with the next iteration, thus avoiding the final else part.

Basically added comments for any novice programmer to understand the code, but I don't think any seasoned / experienced developer will have any difficulty in understanding the code (after all some `if` / `else` s). But as you said, simple code is much better than a very complex one.

Let me know if you still completely prefer the **exact** same way as shown in the above sample link.

PS: the quote, `always write your code as if the person who is going to maintain it, is a serial killer who knows your house`, still gives me a chill... (=

Reply:
> @jarrpa , I have removed the funneling of errors, but I didn't feel the final `else` very compelling in the above example you quoted. Added a `continue` to go on with the next iteration, thus avoiding the final else part.

Removing the final else makes sense in this case, yes.

> PS: the quote, `always write your code as if the person who is going to maintain it, is a serial killer who knows your house`, still gives me a chill... (=

😂😂😂